crates/partition-store/src/vqueue_table/waiting_reader.rs

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2023 - 2025 Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use anyhow::Context;
+use rocksdb::DBRawIteratorWithThreadMode;
+
+use restate_storage_api::StorageError;
+use restate_storage_api::vqueue_table::{EntryCard, Stage, VQueueCursor};
+use restate_types::vqueue::VQueueId;
+
+use crate::PartitionDb;
+use crate::keys::{TableKey, TableKeyPrefix};
+use crate::vqueue_table::InboxKey;
+
+pub struct VQueueWaitingReader {
+    it: DBRawIteratorWithThreadMode<'static, rocksdb::DB>,
+}
+
+impl VQueueWaitingReader {
+    pub(crate) fn new(db: &PartitionDb, qid: &VQueueId) -> Self {
+        let mut readopts = rocksdb::ReadOptions::default();
+        readopts.set_async_io(true);
+        // this is not the place to be concerned about corruption, we favor speed
+        // over safety for this particular use-case.
+        readopts.set_verify_checksums(false);
+        readopts.set_tailing(true);
```

Comment:
Is it a problem that the RocksDB documentation says?

> Not all new data is guaranteed to be available to a tailing iterator. Seek() or SeekToFirst() on a tailing iterator can be thought of as creating an implicit snapshot -- anything written after it may, but is not guaranteed to be seen.